File: 0015-3sum/0015-3sum.swift
// squeeze technique pointers

class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        var copiedNums: [Int] = nums
        copiedNums.sort()
        let length: Int = copiedNums.count
        var answer :[[Int]] = []

        for i: Int in 0..<length{
            if(copiedNums[i] > 0){
                break
            }
            
            // Skip duplicates for the first number
            if i > 0 && copiedNums[i] == copiedNums[i - 1]{
                continue
            }

            var low: Int = i+1
            var high: Int = length - 1

            while low < high {
                let sum: Int = copiedNums[i] + copiedNums[low] + copiedNums[high]

                if sum == 0 {
                    answer.append([copiedNums[i], copiedNums[low], copiedNums[high]])
            
                    // Skip duplicates for the second number
                    while low < high && copiedNums[low] == copiedNums[low + 1]{
                        low += 1
                    }

                    // skip duplicated for the third number
                    while low < high && copiedNums[high] == copiedNums[high - 1]{
                        high -= 1
                    }

                    // move pointers 
                    low += 1
                    high -= 1
                }
                else if sum < 0{
                    low += 1
                }else{
                    high -= 1
                }
                }
            }
            return answer;
        }
    }

File: 0015-3sum/0015-3sum.py
# Input: nums = [-1,0,1,2,-1,-4]
# Output: [[-1,-1,2],[-1,0,1]]

class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        # [-4, -1, -1, 0, 1, 2]
        nums.sort()
        n = len(nums)
        answers = []
        for i in range(n):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i-1]:
                continue
            low = i + 1
            hig = n - 1
            while(low < hig):
                sum = nums[low] + nums[hig] + nums[i]
                if sum == 0:
                    answers.append([nums[low], nums[i], nums[hig]])
                    low += 1
                    hig -= 1
                    while low < hig and nums[low] == nums[low - 1]:
                        low += 1
                    while low < hig and nums[hig] == nums[hig + 1]:
                        hig -= 1
                elif sum < 0 :
                    low += 1
                elif sum > 0:
                    hig -= 1
            
        return answers

print(Solution().threeSum([-1,0,1,2,-1,-4]))